numpy/fft/pocketfft.py

```diff
@@ -173,7 +173,7 @@ def fft(a, n=None, axis=-1, norm=None):
 
     """
 
-    a = asarray(a)
+    a = asarray(a).astype(complex, copy=False)
```

Comment:
To get the dtype more directly just do `asarray(a, dtype=complex)`

Reply:
Thanks, will adjust soon!

Reply:
Do we actually need to cast to `complex` here if the data is already `float64`? It would be worth enumerating in the issue description exactly which types currently fail, so we can work out which casts are actually necessary.

Reply:
Does pocketfft support `float32` without casting here?

Reply:
https://github.com/scipy/scipy/blob/v1.2.0/scipy/interpolate/_bsplines.py#L33

This or similar incantation occurs several times in scipy, and every time the sentiment is "annoyingly, there's no numpy function for it".

Reply:
I think I'm only now beginning to understand the real problem here ...

What exactly does 
```
TypeError: Cannot cast array data from dtype('float128') to dtype('complex128') according to the rule 'safe'
```
mean? So far I thought the code was complaining that it couldn't convert a float type to a complex one, but actually it seems the be about the loss of precision.

If that assumption is correct, we only have to do a cast if the input hast 128bit precision. Everything else will be handled automatically by the existing code.

Is there a way to allow "unsafe" casting in `PyArray_ContiguousFromObject()` and `PyArray_CopyFromObject()`?

Reply:
`complex128` has `float64` real and imaginary, hence it cannot be safely cast indeed. Passing on `astype(..., casting='unsafe')` should solve it.

Sorry, really should have thought of that before...

Reply:
The problem is that I currently do all the hard work (making sure that the input is contiguous and of the proper type) in the C part. If I cannot ask for unsafe casting there, this may lead to duplicated work, which I'd like to avoid. But if the C interface does not have the facility, I'll move this to Python.